src/decode.ts

```diff
@@ -0,0 +1,79 @@
+import { Uint8ArrayList } from 'uint8arraylist'
+import errcode from 'err-code'
+import { FrameHeader, FrameType, HEADER_LENGTH, YAMUX_VERSION } from './frame.js'
+import { ERR_DECODE_INVALID_VERSION } from './constants.js'
+import type { Source } from 'it-stream-types'
+
+/**
+ * Decode a header from the front of a buffer
+ *
+ * @param buffer - Assumed to have enough bytes for a header
+ */
+export function decodeHeader (buffer: Uint8ArrayList): FrameHeader {
+  if (buffer.get(0) !== YAMUX_VERSION) {
+    throw errcode(new Error('Invalid frame version'), ERR_DECODE_INVALID_VERSION)
+  }
+  return {
+    type: buffer.getUint8(1),
+    flag: buffer.getUint16(2, false),
+    streamID: buffer.getUint32(4, false),
+    length: buffer.getUint32(8, false)
+  }
+}
+
+export class Decoder {
+  /** Buffer for in-progress frames */
+  private readonly buffer: Uint8ArrayList
+  private readonly source: Source<Uint8Array>
+
+  constructor (source: Source<Uint8Array>) {
+    this.source = source
+    this.buffer = new Uint8ArrayList()
+  }
+
+  async * emitFrames (): AsyncGenerator<{header: FrameHeader, readData?: () => Promise<Uint8Array>}> {
+    for await (const chunk of this.source) {
+      this.buffer.append(chunk)
+
+      while (true) {
```

Comment:
The while loop is there in case the buffer has enough data for multiple frames / headers, eg if a chunk contains multiple frames. We need to keep consuming bytes until we can't anymore. Otherwise we can get in some situation where the buffer is never fully consumed.
Probably worth commenting it better